src/Mafia/Cabal/Package.hs

```diff
@@ -67,6 +68,27 @@ getCabalFile dir = do
     _ : _ ->
       left $ CabalMultipleFilesFound dir xs
 
+-- | Given a file or directory path, try to find the .cabal file by searching
+--   its parent directories.
+findCabalRoot :: MonadIO m => Path -> m (Maybe Directory)
+findCabalRoot path0 = do
+  path <- canonicalizePath path0
+  entries <- tryGetDirectoryContents path
+  if any (\x -> extension ".cabal" x && x /= ".cabal" && not (T.isSuffixOf "-test.cabal" x)) entries then
```

Comment:
yeah I really am :/ although this is not new for quick/ghci/watch, we already bake in our standard source dir locations:
```hs
standardSourceDirs :: [Directory]
standardSourceDirs =
  ["src", "test", "gen", "dist/build", "dist/build/autogen"]
```